linera-core/src/worker.rs

```diff
@@ -743,28 +743,22 @@ where
     ) -> Result<(), WorkerError> {
         let required_locations = block.bytecode_locations();
         // Find all certificates containing bytecode used when executing this block.
-        let mut required_hashes: HashSet<_> = required_locations
-            .keys()
-            .map(|bytecode_location| bytecode_location.certificate_hash)
-            .collect();
+        let mut required_locations_left: HashMap<CryptoHash, &BytecodeLocation> =
+            required_locations
+                .keys()
+                .map(|bytecode_location| (bytecode_location.certificate_hash, bytecode_location))
```

Comment:
Minor optimization: you can avoid the `location.to_owned()` further down by taking ownership of the locations here (and you can omit the type signature for the generic type parameters):

```suggestion
        let mut required_locations_left: HashMap<_, _> =
            required_locations
                .into_keys()
                .map(|bytecode_location| (bytecode_location.certificate_hash, bytecode_location))
```

Reply:
Ah, `to_owned` clones, but consuming iterators just move the values out, I see. Sounds good!